src/hooks/useSolutionModal/index.tsx

```diff
@@ -0,0 +1,132 @@
+import { Modal } from "@mui/material";
+import { useState } from "react";
+import { PROBLEM_DIFFICULTY_COLOR } from "../../constants/difficulty";
+import { LANGUAGE_COMPILE_MAPPING } from "../../constants/language";
+import CodeBlock from "./CodeBlock";
+import "./styles.css";
+import TabGroup from "./TabGroup";
+
+// hook that returns a function to open the modal, and the modal component
+export default function useSolutionModal() {
+  type LanguageToSolutionMap = Record<string, SolutionAttributes>;
+
+  const [modalOpen, setModalOpen] = useState(false);
+  const [problem, setProblem] = useState<ProblemAttributes | null>(null);
+  const [solutions, setSolutions] = useState<LanguageToSolutionMap>({});
+
+  const solutionLanguages = Object.keys(solutions);
+  const solutionTabs = solutionLanguages.map((language) => {
+    const solutionDetails = solutions[language];
+    const runtime = solutionDetails.runtime_ms;
+    const memory = solutionDetails.memory_usage_mb;
+
+    const hasRuntime = runtime !== null;
+    const hasMemoryUsage = memory !== null;
+
+    return {
+      label: language,
+      content: (
+        <div className="solution-code">
+          <div style={{ padding: "1rem" }}>
+            {hasRuntime && hasMemoryUsage && (
+              <>
+                Runtime: {runtime} ms <br />
+                Memory Usage: {memory} MB
+              </>
+            )}
+            {hasRuntime && !hasMemoryUsage && (
+              <>
+                Runtime: {runtime} ms <br />
+                Memory Usage: {0}B
+              </>
+            )}
+            {!hasRuntime && !hasMemoryUsage && (
+              <div style={{ color: "#FF4500" }}>
+                {solutionDetails.error ?? "Wrong Answer"}
+              </div>
+            )}
+          </div>
+          <CodeBlock
+            code={solutions![language].solution_code}
+            language={LANGUAGE_COMPILE_MAPPING[language]}
+          />
+        </div>
+      ),
+    };
+  });
+
+  function handleOpenSolutionModel(
+    problem: ProblemAttributes,
+    solutions: SolutionAttributes[]
+  ) {
+    const languageSolutionMapping: LanguageToSolutionMap = solutions.reduce(
+      (mapping, solution) => {
+        mapping[solution.solution_language] = solution;
+        return mapping;
+      },
+      {} as LanguageToSolutionMap
+    );
+    setProblem(problem);
+    setSolutions(languageSolutionMapping);
+    setModalOpen(true);
+  }
+
+  function handleCloseSolutionModel() {
+    setProblem(null);
+    setSolutions({});
+    setModalOpen(false);
+  }
+
+  function SolutionModal(): JSX.Element {
+    return (
+      <Modal
+        open={modalOpen}
```

Comment:
Personally I like rendering modals dynamically rather than this variable bc you create and destroy new objects every render rather than managing the state of 1 modal

Reply:
Like modalState === enum.modal && <modal>

Reply:
nah id win